### PR TITLE
Allow setting block time for off-chain testing

### PR DESCRIPTION
**Pull request by CrommVardek**
Adding a `Engine::set_block_timestamp` function to solve issue #1713 

Added a test accordingly.

### CODE REVIEW

File: crates/env/src/engine/off_chain/test_api.rs [hunk 1]
```diff
@@ -240,6 +240,16 @@ where
     })
 }
 
+/// Sets block timestamp for the next advance block.
```
Comment (cmichi) [inline crates/env/src/engine/off_chain/test_api.rs:243]:
```suggestion
/// Sets the block timestamp for the next [`advance_block`] invocation.
```

File: crates/engine/src/exec_context.rs [hunk 2]
```diff
@@ -64,6 +64,11 @@ impl ExecContext {
     pub fn reset(&mut self) {
         *self = Default::default();
     }
+
+    /// Set the execution context block timestamp
```
Comment (cmichi) [inline crates/engine/src/exec_context.rs:68]:
```suggestion
    /// Set the execution context block timestamp.
```

File: crates/engine/src/test_api.rs [hunk 3]
```diff
@@ -257,6 +258,10 @@ impl Engine {
     pub fn set_value_transferred(&mut self, value: Balance) {
         self.exec_context.value_transferred = value;
     }
+
+    pub fn set_block_timestamp(&mut self, new_block_timestamp: BlockTimestamp) {
```
Comment (cmichi) [inline crates/engine/src/test_api.rs:262]:
```suggestion
    /// Set the block timestamp for the execution context.
    pub fn set_block_timestamp(&mut self, new_block_timestamp: BlockTimestamp) {
```

File: crates/engine/src/tests.rs [hunk 4]
```diff
@@ -272,3 +272,20 @@ fn ecdsa_recovery_with_secp256k1_crate() {
...
+#[test]
+fn setting_getting_block_timestamp() {
+    // given
+    let mut engine = Engine::new();
+    let new_block_timestamp: u64 = 1000;
+    let output = &mut &mut get_buffer()[..];
+
+    // when
+    engine.set_block_timestamp(new_block_timestamp);
+    engine.block_timestamp(output);
```
Comment (cmichi) [inline crates/engine/src/tests.rs:285]: The test is pretty good and clearly readable. One thing: could you include testing `advance_block` in the test?

**Review by cmichi — COMMENTED** [review]
The PR is already pretty good, just some nits. 

Could you also update the file `CHANGELOG.md`? There is an `[Unreleased]` section and you can add it under a `## Added` heading there.

File: crates/engine/src/tests.rs (hunk 4 above)
Comment (CrommVardek) [reply crates/engine/src/tests.rs:286]: Essentially, one might want to use the off_chain engine that way : 

```rust 
engine.set_block_time(new_block_time);
engine.advance_block();
engine.block_timestamp(); // Should return the previous block_timestamp + new_block_time
```

"Incomplete" might not be the best word here, but rather, "could have a complementary implementation".

**Review by cmichi — APPROVED** [review]
Thank you, @CrommVardek !

Hope to see more PR's of you :-).

**Pull request merged**